example/main.mo

```diff
@@ -0,0 +1,255 @@
+import Array "mo:base/Array";
+import Error "mo:base/Error";
+import Int "mo:base/Int";
+import Prim "mo:prim";
+import Principal "mo:base/Principal";
+import Text "mo:base/Text";
+import Timer "mo:base/Timer";
+import Vec "mo:vector";
+
+import ICRC1 "../src/ICRC1";
+import TokenHandler "../src";
+
+actor class Example() = self {
+
+  stable var assetsData : Vec.Vector<StableAssetInfo> = Vec.new();
+
+  type AssetInfo = {
+    ledgerPrincipal : Principal;
+    handler : TokenHandler.TokenHandler;
+  };
+  type StableAssetInfo = {
+    ledgerPrincipal : Principal;
+    handler : TokenHandler.StableData;
+  };
+
+  type TokenInfo = {
+    min_deposit : Nat;
+    min_withdrawal : Nat;
+    deposit_fee : Nat;
+    withdrawal_fee : Nat;
+  };
+
+  type NotifyResult = {
+    #Ok : {
+      deposit_inc : Nat;
+      credit_inc : Nat;
+    };
+    #Err : {
+      #CallLedgerError : Text;
+      #NotAvailable : Text;
+    };
+  };
+
+  type WithdrawResult = {
+    #Ok : {
+      txid : Nat;
+      amount : Nat;
+    };
+    #Err : {
+      #CallLedgerError : Text;
+      #InsufficientCredit;
+      #AmountBelowMinimum;
+    };
+  };
+
+  let JOURNAL_SIZE = 1024;
+
+  var initialized : Bool = false;
+  var assets : Vec.Vector<AssetInfo> = Vec.new();
+
+  private func assertInitialized() = if (not initialized) {
+    Prim.trap("Not initialized");
+  };
+
+  public shared func init() : async () {
+    assert not initialized;
+    assets := Vec.map<StableAssetInfo, AssetInfo>(
+      assetsData,
+      func(x) {
+        let r = (actor (Principal.toText(x.ledgerPrincipal)) : ICRC1.ICRC1Ledger)
+        |> {
+          balance_of = _.icrc1_balance_of;
+          fee = _.icrc1_fee;
+          transfer = _.icrc1_transfer;
+          transfer_from = _.icrc2_transfer_from;
+        }
+        |> {
+          ledgerPrincipal = x.ledgerPrincipal;
+          handler = TokenHandler.TokenHandler(_, Principal.fromActor(self), JOURNAL_SIZE, 0, true);
+        };
```

Comment:
It looks like we need some convenience functions to reduce this boiler plate code for the developer.

For example, we could provide a convenience function in ICRC1.mo which takes either a principal or the actor and produces the type LedgerAPI which can then be passed as the first argument to the TokenHandler constructor.

Or we need a second constructor in lib.mo which wraps around TokenHandler and which takes a principal instead of type LedgerAPI as first argument and does the rest internally.